spirv_msl.cpp

```diff
@@ -139,6 +141,53 @@ void CompilerMSL::build_implicit_builtins()
 			builtin_sample_id_id = var_id;
 		}
 	}
+
+	if (msl_options.swizzle_texture_samples && has_sampled_images)
+	{
+		uint32_t offset = increase_bound_by(5);
+		uint32_t type_id = offset;
+		uint32_t type_arr_id = offset + 1;
+		uint32_t struct_id = offset + 2;
+		uint32_t struct_ptr_id = offset + 3;
+		uint32_t var_id = offset + 4;
+
+		// Create a buffer to hold the swizzle constants.
+		SPIRType uint_type;
+		uint_type.basetype = SPIRType::UInt;
+		uint_type.width = 32;
+		set<SPIRType>(type_id, uint_type);
+
+		SPIRType uint_type_arr = uint_type;
+		uint_type_arr.array.push_back(0);
+		uint_type_arr.array_size_literal.push_back(true);
+		uint_type_arr.parent_type = type_id;
+		set<SPIRType>(type_arr_id, uint_type_arr);
+		set_decoration(type_arr_id, DecorationArrayStride, 4);
+
+		SPIRType struct_type;
+		struct_type.basetype = SPIRType::Struct;
+		struct_type.member_types.push_back(type_arr_id);
+		auto &type = set<SPIRType>(struct_id, struct_type);
+		type.self = struct_id;
+		set_decoration(struct_id, DecorationBlock);
+		set_name(struct_id, "spvAux");
+		set_member_name(struct_id, 0, "swizzleConst");
+		set_member_decoration(struct_id, 0, DecorationOffset, 0);
+
+		SPIRType struct_type_ptr = struct_type;
+		struct_type_ptr.pointer = true;
+		struct_type_ptr.parent_type = struct_id;
+		struct_type_ptr.storage = StorageClassUniform;
+		auto &ptr_type = set<SPIRType>(struct_ptr_id, struct_type_ptr);
+		ptr_type.self = struct_id;
+
+		set<SPIRVariable>(var_id, struct_ptr_id, StorageClassUniform);
+		set_name(var_id, "spvAuxBuffer");
+		// This should never match anything.
+		set_decoration(var_id, DecorationDescriptorSet, 0xFFFFFFFE);
```

Comment:
Why 0xfffffffe in particular? Does it need a descriptor set decoration?

Reply:
It needs a unique descriptor set so it doesn't collide with any descriptor sets specified by the client, or with the special push constant descriptor set.

Reply:
Fair enough.